core/systems/disks/steam_removable_media_manager.gd

```diff
@@ -0,0 +1,241 @@
+
+# TODO: Fix this workaround to allow similtaneous operations. Currently more than
+# one operation will reset the drive nodes and permit undefined behavior.
+var block_operations: bool = false
+
+# Polkit Script paths
+const SRM_FORMAT_MEDIA = "/usr/bin/shadowblip/format-media"
+const SRM_INIT_MEDIA = "/usr/bin/shadowblip/init-media"
+const STEAMOS_RETRIGGER_AUTOMOUNTS = "/usr/bin/steamos-polkit-helpers/steamos-retrigger-automounts"
+const STEAMOS_FORMAT_SDCARD = "/usr/bin/steamos-polkit-helpers/steamos-format-sdcard"
+const STEAMOS_TRIM_DEVICES = "/usr/bin/steamos-polkit-helpers/steamos-trim-devices"
+const BLOCK_PREFIX = "/org/freedesktop/UDisks2/block_devices"
+const SDCARD_PATH = "/dev/mmcblk0"
+
+const protected_mounts = [
+	"/",
+	"/boot",
+	"/boot/efi",
+	"/efi",
+	"/frzr_root",
+	"/frzr_root/boot",
+	"/home",
+	"/var",
+	"/var/cache",
+	"/var/log",
+	]
+
+var format_capable: bool = false
+var init_capable: bool = false
+var format_sd_capable: bool = false
+var retrigger_capable: bool = false
+var trim_capable: bool = false
+
+
+signal drives_updated(device_tree: Array[BlockDevice])
+
+## Hierarchy of drive -> block device -> partitions -> mount points
+## Used to display device information to the user.
+var device_tree: Array[BlockDevice]
+
+func _init() -> void:
+
+	# Check the required system files exist for steam_removable_media
+	if FileAccess.file_exists(SRM_FORMAT_MEDIA):
+		format_capable = true
+	if FileAccess.file_exists(SRM_INIT_MEDIA):
+		init_capable = true
+	if FileAccess.file_exists(STEAMOS_RETRIGGER_AUTOMOUNTS):
+		retrigger_capable = true
+
+	# These scripts are hard coded to mmcblk0
+	if FileAccess.file_exists(SDCARD_PATH):
+		if FileAccess.file_exists(STEAMOS_FORMAT_SDCARD):
+			format_sd_capable = true
+		if FileAccess.file_exists(STEAMOS_TRIM_DEVICES):
+			trim_capable = true
+
+	logger.debug("format_capable: " + str(format_capable))
+	logger.debug("init_capable: " + str(init_capable))
+	logger.debug("format_sd_capable: " + str(format_sd_capable))
+	logger.debug("retrigger_capable: " + str(retrigger_capable))
+	logger.debug("trim_capable: " + str(trim_capable))
+
+	# Get the current drives and ensure they are always up to date
+	udisks2.devices_updated.connect(update_drives)
+	update_drives()
+
+
+## Rebuilds the device_tree with the current device list
+func update_drives(devices: Array[BlockDevice] = []):
+	if devices == []:
+		logger.debug("Got empty array, grabbing devices from udisks2")
+		devices = udisks2.get_devices()
+	self.device_tree = []
+	for block in devices:
+		if not _block_device_has_protected_mount(block):
+			logger.debug(block.dbus_path, "has no protected mounts.")
+			device_tree.append(block)
+	self.drives_updated.emit(self.device_tree)
+
+
+## Calls the SteamRemovableMedia format-media script to format a drive as EXT4
+## and intialize it as a steam library
+func format_drive(device: BlockDevice) -> void:
+	if not format_capable:
+		logger.error("System is not format capable")
+		return
```

Comment:
We should return an error code for this function, so callers can determine if a format failed or not.